File: solution/106. Construct Binary Tree from Inorder and Postorder Traversal/106.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        unordered_map<int, int> m;
        for (int i = 0; i < inorder.size(); i++)
            m[inorder[i]] = i;
        
        int pRoot = postorder.size() - 1;
        return buildTree(m, 0, inorder.size()-1, postorder, pRoot);
    }
    
private:
    TreeNode* buildTree(unordered_map<int, int>& m, int inStart, int inEnd, vector<int>& postorder, int& pRoot) {
        if (inStart > inEnd) return nullptr;
        
        TreeNode* root = new TreeNode(postorder[pRoot--]);

        root->right = buildTree(m, m[root->val]+1, inEnd, postorder, pRoot);
        root->left = buildTree(m, inStart, m[root->val]-1, postorder, pRoot);

        return root;
    }
};

File: solution/106. Construct Binary Tree from Inorder and Postorder Traversal/106.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int postIndex;
    private int[] inorder, postorder;
    private HashMap<Integer, Integer> m;
    
    public TreeNode buildTree(int[] inorder, int[] postorder) {
        this.inorder = inorder;
        this.postorder = postorder;
        m = new HashMap<Integer, Integer>();
        for (int i = 0; i < inorder.length; ++i)
            m.put(inorder[i], i);
        
        postIndex = postorder.length - 1;
        return buildTree(0, inorder.length - 1);
    }
    
    private TreeNode buildTree(int inStart, int inEnd) {
        if (inStart > inEnd) return null;
        int rootVal = postorder[postIndex--];
        TreeNode root = new TreeNode(rootVal);
        root.right = buildTree(m.get(rootVal) + 1, inEnd);
        root.left = buildTree(inStart, m.get(rootVal) - 1);
        return root;
    }
}
